contracts/zero-ex/contracts/src/features/SimpleFunctionRegistry.sol

```diff
@@ -0,0 +1,184 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "../interfaces/IFeature.sol";
+import "../interfaces/ISimpleFunctionRegistry.sol";
+import "../interfaces/IZeroExBootstrapper.sol";
+import "../fixins/FixinOwnable.sol";
+import "../storage/LibProxyStorage.sol";
+import "../storage/LibSimpleFunctionRegistryStorage.sol";
+import "../errors/LibSimpleFunctionRegistryRichErrors.sol";
+
+
+/// @dev Basic registry management features.
+contract SimpleFunctionRegistry is
+    IFeature,
+    ISimpleFunctionRegistry,
+    IZeroExBootstrapper,
+    FixinOwnable
+{
+    // solhint-disable const-name-snakecase
+
+    /// @dev Name of this feature.
+    string constant public override FEATURE_NAME = "SimpleFunctionRegistry";
+    /// @dev Version of this feature.
+    uint256 constant public override FEATURE_VERSION = (1 << 64) | (0 << 32) | (0);
+
+    /// @dev Initializes the feature implementation registry.
+    /// @param impl The actual address of this feature contract.
+    function bootstrap(address impl) external override {
+        // Register the registration functions (inception vibes).
+        _extend(this.extend.selector, impl);
+        _extend(this.extendSelf.selector, impl);
+        // Register the rollback function.
+        _extend(this.rollback.selector, impl);
+        // Register getters.
+        _extend(this.getRollbackLength.selector, impl);
+        _extend(this.getRollbackEntryAtIndex.selector, impl);
+    }
+
+    /// @dev Roll back to a prior implementation of a function.
+    ///      Only directly callable by an authority.
+    /// @param selector The function selector.
+    /// @param targetImpl The address of an older implementation of the function.
+    function rollback(bytes4 selector, address targetImpl)
+        external
+        override
+        onlyOwner
+    {
+        (
+            LibSimpleFunctionRegistryStorage.Storage storage stor,
+            LibProxyStorage.Storage storage proxyStor
+        ) = _getStorages();
+
+        address currentImpl = proxyStor.impls[selector];
+        if (currentImpl == targetImpl) {
+            // Do nothing if already at targetImpl.
+            return;
+        }
+        // Walk history backwards until we find the target implementation.
+        address[] storage history = stor.implHistory[selector];
+        uint256 i = history.length;
+        for (; i > 0; --i) {
+            address impl = history[i - 1];
+            history.pop();
+            if (impl == targetImpl) {
+                break;
+            }
+        }
+        if (i == 0) {
```

Comment:
Wouldn't this get triggered below when we try to rollback `extendSelf`?

Reply:
No, because `extendSelf` gets added through `_extend()`, which pushes the old implementation (null) to the stack, so it has a history at that point.